sklearn/preprocessing/_discretization.py

```diff
@@ -236,20 +250,32 @@ def fit(self, X, y=None):
 
         return self
 
-    def _validate_n_bins(self, n_features):
+    def _validate_n_bins(self, n_features, n_samples):
         """Returns n_bins_, the number of bins per feature.
         """
-        orig_bins = self.n_bins
+        orig_bins = self._n_bins
+        if isinstance(orig_bins, str):
+            if orig_bins == 'auto':
+                # calculate number of bins with Sturges rule
+                orig_bins = max(int(np.ceil(np.log2(n_samples) + 1.)), 2)
```

Comment:
Is there an advantage of using the `'auto'` behavior defined by [np.histogram_bin_edges](https://numpy.org/devdocs/reference/generated/numpy.histogram_bin_edges.html#numpy.histogram_bin_edges)? In their case, they take the maximum of `'sturges'` and `'fd'`.

Reply:
Thank you for your detailed thoughts!

If one were planning to discretize their feature, I do not think we would require them to scale their feature. This means that, for benchmarking, I would assume that the data is not scaled when comparing the different techniques.

>  tested Rice, Sqrt and Sturges rule on some toy datasets and with different type of models

Can you provide the code for these benchmarks? When implementing a feature without much literature backing it, I tend to try to see what kind of benchmarks we can show to see when this is useful. There looks to be benchmarks here:

- https://ai.stanford.edu/~ronnyk/disc.pdf
- https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2656082/
 
but they do not use `sturges` or `fd`. I could be missing a paper somewhere that actually looks into `sturges` or `fd` for ML.

> And also we will need to calculate iqr for every feature in dataset and determine n_bins for every feature, instead of unified approach.

Codewise, it would be okay, since the IQR computation is `np.percentile(X, q=[0.25, 0.75], axis=0)`, and we already support `n_bins` as an array-like.

Also looks like @amueller and @janvanrijn wrote a paper on this here: https://arxiv.org/pdf/2007.07588.pdf where they wrote on page 3:

> The Freedman-Diaconis rule is known to be re- silient to outliers. An alternative of this rule is Sturges formula [14]. The automatic histogram binning, as implemented in the python library numpy [10], uses the maximum of the Freedman-Diaconis rule and Sturges formula. However, Sturges formula assumes that the data follows a Gaussian distribution. Since we have no reason to assume that this holds for our data, we only use the Freedman- Diaconis rule.

Reply:
My bad. FD rule suggest not the number of bins but the width of bins. We cannot compare the two directly.

1. In original paper FD assumed that the number of bins should be proportional to cbrt(N) ([cbrt(6*gamma^(-1/3)*N), to be precise](https://bayes.wustl.edu/Manual/FreedmanDiaconis1_1981.pdf)), which is essentially Rice rule. If you really want to use Rice rule, than we don't even need to compare it to Sturges, since it will always be bigger (and grow faster) and we won't need to calculate IQR.
2. If we really want to compare Sturges and FD than we will need to calculate IQR `(np.diff(np.nanquantile(X, [0.25, 0.75], axis=0), axis=0)` or `stats.iqr(X, axis=0, nan_policy='omit')`, so we will need to refactor code (pass X to `_validate_bins` function at least). On top of that to convert bin width to number of bins we will need to calculate range `np.ptp(X, axis=0) ` which will also introduce dependence on 'Uniform' strategy, while `KBinsDiscretizer` also has 'Quantile' and 'Kmeans' (and I guess Spline strategy could be next). 'Uniform' strategy makes a lot of sense for plotting the data (which is `np.histogram` is for) because it saves the form of distribution, but not always used in ML (every GBM and GPU RF uses 'Quantile' strategy for example). On top of that FD will almost always give bigger numbers than Sturges, so we don't even need to compare to it, just `max(FD, 2)` would be enough.
3. IQR is robust to outliers, yes, but that's not what I said. I said it relies on normality of the data, or at least to not so skewed data. In original paper, they optimized L2 functional, which is useful only on symmetric distribution. Optimizing L2 on Exponentially distributed data won't give the expected results, for instance.
4. Those papers compare different discretization strategies. In `KBinsDiscretizer` we have 'Uniform' (equal width), 'Quantile' (equal frequency) and 'KMeans'. Can you please elaborate how **discretization strategy** is connected to **selecting number of bins**?
5. The code was pretty simple: `make_classification` and `make_regression` to create data, `KBinsDiscretizer` + `LogisticRegression`/`RandomForestClassifier` or `Ridge`/`RandomForestRegressor` in `cross_val_score` function with R2 and ROC AUC metrics.

Reply:
>  Can you please elaborate how discretization strategy is connected to selecting number of bins?

I was mistaken and the papers I listed does not apply.

On the surface this looks like a simple feature to sklearn, but we still have an [inclusion criteria](https://scikit-learn.org/stable/faq.html#what-are-the-inclusion-criteria-for-new-algorithms). The only source I have that directly speaks about struges and fd in the ML context is the https://arxiv.org/pdf/2007.07588.pdf paper, where they explicitly use fd.

To try to move this forward, I suggest:

1. Changing the name from `'auto'` to `'sturges'`. We can not change the behavior of 'auto' without a deprecation cycle, so being more explicit would be better. (A deprecation cycle consist of giving a warning about changing behavior for 2 releases and then changing the behavior.)
2. Benchmark to compare it to the current default=5 **or** find a paper that uses struges in the ML context.

I suspect that comparing struges to `n_bins=5` would result in struges being better. For simplicity, it would use `make_*` but with different `n_samples` to see how the performance compares between 5 and struges.